include/tvm/ir/function.h

```diff
@@ -63,6 +63,43 @@ enum class CallingConv : int {
    * - Implementation: defined by device runtime(e.g. runtime/cuda)
    */
   kDeviceKernelLaunch = 2,
+  /*!
+   * \brief For functions only called by other functions within the same IR module.
+   *
+   * This calling convention exists to support one PrimFunc calling another PrimFunc
+   * within the same IRModule.
+   *
+   * Overview / Purpose:
+   *
+   *   - This calling convention is intended only for PrimFuncs whose callers reside in
+   *     the same IRModule.  This is the only supported use case.
+   *
+   *   - The details of the calling convention may change frequently as TVM evolves.
+   *     Therefore users are discouraged from attempting to use this calling convention
+   *     outside of the supported use case(s).
+   *
+   * Current mechanics / usage requirements:
+   *
+   *   - A PrimFunc with this calling convention will NOT undergo any of the signature
+   *     transformations provided by the MakePackedAPI pass.
+   *
+   *   - Supported use cases, and their corresponding unit tests, are all expressed as
+   *     TVMScript.
+   *
+   *   - The callsite must use `T.call_extern`.
```

Comment:
Alternatively, perhaps we can clarify it as C calling conv, effectively pass things as they are, which seems to match the behavior of call_extern and current behavior.

Reply:
Thanks for the feedback!

Could you clarify what you mean by "C calling convention" in this context?

I've only heard that term used to describe the mapping of C-language types to (e.g.) x86-64 ABI -compliant function calls.

But in this PR's context, the question (AFAICT) is how we map TVMScript callsite arguments to TVMScript formal parameters.  So I'm not sure how this would relate to a C calling convention.

Reply:
Just want to get some clarity here. That boils down how we map this cross function call in codegen phase.  

- If the call maps into call_extern => caller side LLVM call,  callee side just present the function as LLVM function. Then it is the common calling convention(either known as default per defined by platform)
- The reason why it is related to "C" calling convention is that if the callee side say was implemented as a C function with the same signature(and compile that say through hexagon-llvm, or platform-c), and we compile caller side as it is, it will still work

If the cross function call will be further lowered(and the behavior), then it might be useful to give it a different name. IntraModule might be slightly too generic(since we can also use Packed call intra module), also the caller side might need a different intrinsic(as call_extern implies call into a function with C calling convention). So if we want to state that this is a convention that will be further lowered, just a strawman: (e.g. InternalModuleCall, call_internal)

After reading the context, seems we lean towards later. So the main comment to clarify this is to relate the calling convention as much as possible to existing known ones, or clarify it so it is differentiated from existing known ones(by different intrinsics and name).

Reply:
> as call_extern implies call into a function with C calling convention

To double-check, is this the case for functions that exist within the same module?  That is, because the name resolution in the LLVM codegen [preferentially resolves to names](https://github.com/apache/tvm/blob/main/src/target/llvm/codegen_llvm.cc#L973) within the existing module, my mental model was that the calling convention was only fixed for calls into external functions, or for callees that may be called externally.  Therefore, the internal-only function would be sufficient to modify any matching `call_extern` instances.

> So if we want to state that this is a convention that will be further lowered, just a strawman: (e.g. InternalModuleCall, call_internal)

From the earlier design discussions, this was brought up as a potential later step, for when using `call_extern` becomes too unwieldy, or unable to express the degrees of freedom required by internal calls.  Since it seemed that the changes to `call_extern` were legal with the current semantics, we hadn't moved to make a new builtin for it.

> IntraModule might be slightly too generic(since we can also use Packed call intra module),

I'm not quite seeing the issue with the name here.  The semantics we were picturing was that while any calling convention may be used for internal calls, only the `kIntraModule` calling convention  That is, `kIntraModule` implies an internal function, but not being `kIntraMoudule` does not imply the existence of external calls.  The difference would be that a pass could modify the signature of a `kIntraModule` function (e.g. removing an unused argument from the signature, simultaneously updating all callsites), while a PackedFunc could not be modified without breaking the potential external callsites.

> After reading the context, seems we lean towards later. In that case, it might still be useful to specify what would be the default lowering of InternalModuleCall and call_internal.

The concern was that by specifying that behavior, it would enable assumptions of the final calling convention before it had been selected, mixing the different layers of abstraction.  That is, somebody could erroneously generate a callsite that follows the PackedFunc API, on the assumption that the callee will also be lowered to the PackedFunc API.  By leaving it unspecified, it makes that misunderstanding less likely.

Reply:
Thanks for the discussion.

> The difference would be that a pass could modify the signature of a kIntraModule function (e.g. removing an unused argument from the signature, simultaneously updating all callsites), while a PackedFunc could not be modified without breaking the potential external callsites.

This have to do with linkage which is somewhat parallel to the calling convention. We should be able to modify a signature of a function if it is not being referred from outside (in our current case, no `global_symbol` attribute indicates a private linkage), and we can do such setups independent from the calling convention.


>  it would enable assumptions of the final calling convention before it had been selected, mixing the different layers of abstraction

Indeed this is what we aim to avoid here. That is why normally we try to have pairs of(callsite intrinsic, and callee side calling convention annotations). For example, 

- `InternalModuleCall` (callee side conv) and `call_internal`(caller side intrinsic) would make such pairing clear.

In the current proposal, we are trying to pair (`kIntraModule` and `call_extern`). Where previously in our convention `call_extern` was normally paired with C calling conv(mainly because we also can generate such calls).


BTW, I agree that it is OK to leave default lowering behavior unspecified, as long as the pairing is clear. But in this case, the possible confusion is that we already have lowering behavior for `call_extern` defined by our backend. So only leaving the callee site unspecified could leads to possible point of confusion.

So the main comment here is let us make sure our pairing is clear. Ideally we have a unique set of pairings, e.g. `call_extern` only maps to one callee-site calling conv. Introducing a extra intrinsic to pair with this one, or clarify that this is the only thing that call_extern pairs(in which case perhaps C calling conv is more proper) might be  more clear.

The internal property have things to do with linkage and can be handled separately

Reply:
> Introducing a extra intrinsic

Could we leverage the fact that `call->op` can be `GlobalVar` to explicitly pair with `kIntraModule` function?

For `call_extern` , I believe `call_extern` an intra-module function by name should work in general settings. While except the calling convention issue, it seems also assume the presence of runtime symbol lookup machanism in runtime (like dlsym in glibc), thus may not perfectly match so-called "intra-module" semantic? For example, one may want to compile both \_\_global__ function and \_\_device__ function (maybe kIntraModule) into single cuda module and invoke the \_\_device__ function directly from \_\_global__ function.

Reply:
@wrongtest-intellif : Thanks very much for the thoughtful suggestion!

I don't have a ready answer to your suggestion, which means I have some more homework to do on my side before proceeding with the PR.  Much appreciated!